File: fixtures/java/src/com/acme/order/AbstractOrder.java
package com.acme.order;

/** Shared state for every kind of order. */
public abstract class AbstractOrder implements Discountable {

    protected Money total;

    private final String reference;

    protected AbstractOrder(String reference) {
        this.reference = reference;
        this.total = Money.zero();
    }

    public String reference() {
        return reference;
    }

    @Override
    public Money price() {
        return total;
    }
}

File: fixtures/csharp/src/Acme/Order/AbstractOrder.cs
namespace Acme.Order;

/// <summary>Shared state for every kind of order.</summary>
public abstract class AbstractOrder : IDiscountable
{
    protected Money total;

    private readonly string reference;

    protected AbstractOrder(string reference)
    {
        this.reference = reference;
        total = Money.Zero();
    }

    /// <summary>An auto-property: a value holder with no body of its own.</summary>
    public string Reference => reference;

    public Money Price()
    {
        return total;
    }

    public abstract Money Discount(int percent);
}
